airflow/api/common/experimental/mark_tasks.py

```diff
@@ -0,0 +1,152 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+
+from airflow.jobs import BackfillJob
+from airflow.models import DagRun
+from airflow.operators.subdag_operator import SubDagOperator
+from airflow.settings import Session
+from airflow.utils.state import State
+
+
+def _create_dagruns(dag, execution_dates, state, run_id_template):
+    """
+    Infers from the dates which dag runs need to be created and does so.
+    :param dag:
+    :param execution_dates:
+    :param session:
+    :return:
+    """
+    # find out if we need to create any dag runs
+    drs = DagRun.find(dag_id=dag.dag_id, execution_date=execution_dates)
+    drs_to_create = list(set(execution_dates) - set([dr.execution_date for dr in drs]))
+
+    for date in drs_to_create:
+        dr = dag.create_dagrun(
+            run_id=run_id_template.format(date.isoformat()),
+            execution_date=date,
+            start_date=datetime.now(),
+            external_trigger=False,
+            state=state,
+        )
+        drs.append(dr)
+
+    return drs
+
+
+def set_state(task, execution_date, upstream=False, downstream=False,
+              future=False, past=False, state=State.SUCCESS):
+    """
+    Set the state of task and if needed its relatives. Can set state
+    for future tasks (calculated from execution_date) and retroactively
+    for past tasks. Will verify integrity of past dag runs in order to create
+    tasks that did not exist.
+    :param task: the task from which to work. task.task.dag needs to be set
+    :param upstream: Mark all parents (upstream tasks)
+    :param downstream: Mark all siblings (downstream tasks) of task_id, including SubDags
+    :param future: Mark all future tasks on the interval of the dag up until
+        last execution date.
+    :param past: Retroactively mark all tasks starting from start_date of the DAG
+    :param state: State to which the tasks need to be set
+    :return: list of tasks that has been created and updated
+    """
+    assert isinstance(execution_date, datetime.datetime)
+    execution_date = execution_date.replace(microsecond=0)
+
+    assert task.dag is not None
+    dag = task.dag
+
+    # determine date range of dag runs and tasks to consider
+    end_date = ((dag.latest_execution_date or datetime.datetime.now())
+                if future else execution_date)
+
+    if 'start_date' in dag.default_args:
+        start_date = dag.default_args['start_date']
+    elif dag.start_date:
+        start_date = dag.start_date
+    else:
+        start_date = execution_date
+
+    start_date = execution_date if not past else start_date
```

Comment:
Nit: Would be good to combine this with the if statements above to avoid overwriting previous assignments and simplifying the logic